### DISCUSSION TITLE
Feature: Extract refactoring

### DISCUSSION THREAD

**Discussion by kontomondo** [ionide/ionide-vscode-fsharp#172] [Ideas]
The idea is to extract a selected block of code - a function or a value, and give it a name. In case a block is extracted that uses values that are not available in the target scope, they are introduced as parameters (not a problem if it's decided to extract to the parent scope for simplicity). If extracting a part of the string, the correct concatenation is made. If extracting a lambda function the brackets are dropped and `fun` is removed.

eg. 

``` fsharp
let noSpaces = chars |> Array.filter ((<>) ' ')
```

selecting `((<>) ' ')` and hitting extract asks for the name, and if given a name "`notSpace`", leads to:

``` fsharp
let notSpace = ((<>) ' ')
let noSpaces = chars |> Array.filter notSpace
```

and

``` fsharp
let res = (o.Data.Overloads |> Array.fold (fun acc n -> (n |> Array.toList) @ acc ) []).Head
```

selecting the fold function and hitting extract asks for the name, and if given a name "`collectSignatures`", leads to:

``` fsharp
let collectSignatures acc n = 
    (n |> Array.toList) @ acc 
let res = (o.Data.Overloads |> Array.fold collectSignatures []).Head
```

On a part of a string:

``` fsharp
let bashPath = "/usr/bin/bash"
let zshPath = "/usr/bin/zsh"
```

if "/usr/bin/" is selected, extract is invoked, the name "`prefix`" is given, it results in

``` fsharp
let prefix = "/usr/bin/"
let bashPath = prefix + "bash"
let zshPath = prefix + "zsh"
```

Ideally this would also replace all occurrences of the extracted block. To make it simpler at first, this feature should probably extract to the immediate outer scope.

**Comment by dmannock** [comment]
Hey,
Not to hijack the thread. However, I was after something similar so made a naive version called [vscode-fsharp-refactor](https://marketplace.visualstudio.com/items?itemName=danmannock.vscode-fsharp-refactor)

It currently only helps with the first let binding extraction example you gave but it's early days yet so I'm aiming to make it as useful as possible.

Hope this helps! I appreciate any feedback.

**Comment by forki** [comment]
@dmannock are you planning to contribute it to ionide?

**Comment by forki** [comment]
I ask because the f# community is pretty small and we should always try to work together on stuff. In any case: having more refactorings will always be appreciated.

**Comment by dmannock** [comment]
@forki Yea sure, I'd love too!

It started as a POC to see if it would work easily without using a language server. Since it worked for my use cases, I thought I might as well share it and see if there are any other suggestions/use cases to improve.

I then came across this issue which is labelled as 'requires additional upstream functionality'
- what exactly does this entail?
- and subject to the above, what's the process for getting these features in to ionide? 

I'd currently class my extension as 'experimental' and till it covers more use cases wouldn't want to put in in to ionide :)

Thanks

**Comment by jkone27** [comment]
any updates on this, or is there any way to do extract refactor in 2023  in ionide? using this i get most times "no refactorings available" 
https://marketplace.visualstudio.com/items?itemName=danmannock.vscode-fsharp-refactor

**Reply by baronfel** [reply]
No updates on this, but it would be a very useful contribution. There are a couple ways to push on it:

* implement a lot of the core logic (identifying an expression, identifying the scope at which to insert the expression, etc) in FCS
* write VS and Iondie-specific bindings to that logic

or 

* prototyping that logic in Ionide
* upstreaming to dotnet/fsharp after the functionality is made solid
